CONTRIBUTING_TO_OPENSHIFT_PLUGINS.md

```diff
@@ -0,0 +1,414 @@
+READMEs.
+
+## Constructing your development sandbox
+
+You will need an environment with the following tools installed:
+
+* Maven (the `mvn` command)
+* Git
+* Java (need v8)
+* an IDE or editor of your choosing
+
+The Jenkins open source project already has a bunch of useful links on setting up your development environment, including 
+launching Jenkins from the IDE (though there are pros and cons for developing this way vs. developing against a Jenkins 
+server running in an OpenShift pod).
+
+Here are a few of those links for reference:
+
+* [https://wiki.jenkins.io/display/JENKINS/Extend+Jenkins](https://wiki.jenkins.io/display/JENKINS/Extend+Jenkins) (note, there are many useful child pages under this wiki page)  
+* [https://wiki.jenkins.io/display/JENKINS/Plugin+tutorial](https://wiki.jenkins.io/display/JENKINS/Plugin+tutorial) (one of the key child pages under Extend Jenkins)
+* [https://wiki.jenkins.io/display/JENKINS/Setting+up+Eclipse+to+build+Jenkins](https://wiki.jenkins.io/display/JENKINS/Setting+up+Eclipse+to+build+Jenkins) (has concepts common to other IDEs undoubtedly)
+
+### Basic code development flow
+
+Our plugins are constructed such that if they are running in an OpenShift pod, they can determine how to connect to the 
+associated OpenShift master automatically, and no configuration of the plugin from the Jenkins console is needed.
+
+If you choose to run an external Jenkins server and you would like to test interaction with an OpenShift master, you will need to manually configure the plugin.  See each plugin's README or the OpenShift documentation
+for the specifics.
+
+An example flow when running in an OpenShift pod:
+
+1. Clone this git repository:
+    ```
+    git clone https://github.com/openshift/<plugin in question>-plugin.git
+    ```
+1. In the root of the local repository, run maven 
+    ```
+    cd <plugin in question>-plugin
+    mvn
+    ```
+1. Maven will build target/<plugin name>.hpi  (the Jenkins plugin binary)
+1. Open Jenkins in your browser, log in as an administrator, and navigate as follows:
+1. Manage Jenkins > Manage Plugins.
+1. Select the "Advanced" Tab.
+1. Find the "Upload Plugin" HTML form and click "Browse".
+1. Find the <plugin name>.hpi built in the previous steps.
+1. Submit the file.
+1. Check that Jenkins should be restarted.
+
+### Additional options for updating the plugin in Jenkins (when running in OpenShift)
+
+Aside from updating the plugin via the Jenkins plugin manager, OpenShift's capabilities provide a myriad 
+of other options.  For example:
+
+* Considering the [image creation/update options](https://github.com/openshift/jenkins#adding-plugins-or-updating-existing-plugins)
+available with the OpenShift Jenkins image, one could update the image and either manually recycle the deployment, or 
+* Follow the approach of our extended test (see below), but instantiate the OpenShift templates for Jenkins manually
+* Construct templates such as this one (thanks @csrwng):
```

Comment:
It's hard to parse these 3 bullets and understand how they do/do not relate.  The first bullet says "considering the image creation options available in openshift, one could update the image and either manually recycle the deployment or"  (or what?  or do what is described in the second bullet?  what is described in either the second or third bullet?  it's not obvious to me that either of the next two bullets are direct alternatives to manually recycling the deployment).

it would also help to give a little context to the template method (what does the template do? how does it update the plugins?  how do you iterate with it?)

Frankly i'd be somewhat inclined to just simplify this(the 3 bullets) down to "build a new image containing your plugin and point your jenkins deployment to the image".  (it's no different from developing any other application on openshift at that point).

But if you want to clean up the three approaches i'm fine with that too, I just think they need more context/hand-holding if they're going to stay here, otherwise i'm afraid they just add confusion.